solrmonitor/solrmonitor.go

```diff
@@ -311,25 +311,41 @@ func (c *SolrMonitor) shouldWatchChildren(path string) bool {
 }
 
 func (c *SolrMonitor) dataChanged(path string, data string, version int32) error {
-	if !strings.HasPrefix(path, c.solrRoot+"/collections/") || !strings.HasSuffix(path, "/state.json") {
-		return fmt.Errorf("unknown dataChanged: %s", path)
-	}
-	coll := c.getCollFromPath(path)
-	if coll != nil {
-		coll.setData(data, version)
-		c.callSolrListener(coll)
-		if coll.isPRSEnabled() {
-			coll.startMonitoringReplicaStatus()
+	collectionsPrefix := c.solrRoot + "/collections/"
+	if strings.HasPrefix(path, collectionsPrefix) {
+		// we will always need the collection
+		coll := c.getCollFromPath(path)
+		if strings.HasSuffix(path, "/state.json") {
+			// common state.json change
+			if coll != nil {
+				state := coll.setStateData(data, version)
+				c.callSolrListener(coll.name, state)
+				if coll.isPRSEnabled() {
+					coll.startMonitoringReplicaStatus()
+				}
+			}
+			return nil
+		} else if trimmed := strings.TrimPrefix(path, collectionsPrefix); !strings.Contains(trimmed, "/") {
+			// less common (usually just initialization) collection change
+			if coll != nil && coll.name == trimmed {
```

Comment:
What happens if `coll !=nil and coll.name != trimmed` ? That seems like an error condition?

This trimming code looks weird. Shouldn't we be leaning on `c.getCollFromPath` to do this work for us?

Reply:
> What happens if `coll !=nil and coll.name != trimmed` ? That seems like an error condition?

Yeah, I can update to add an error condition here! Realistically shouldn't be possible.

> This trimming code looks weird. Shouldn't we be leaning on `c.getCollFromPath` to do this work for us?

Only thing with `getCollFromPath` is that it doesn't technically enforce a single path component where the org id is. My goal here was to only accept paths that were collection prefix + a single path component. But how about I have `getCollFromPath` give the prefix and suffix it found? Then we can do easier checks.